Add matter attributes

jvmahon opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Adds entities to set or report on several additional matter attributes:

**Basic Information Cluster**
LocalConfigDisabled (added as a Select entity)

**DoorLock Cluster:**
AutoRelockTimer (added as a Number entity)
SoundVolume (added as a Select entity)
EnableOneTouchLocking (added as a Select entity)
MinPINCodeLength (added as a diagnostics sensor).
MaxPINCodeLength (added as a diagnostics sensor).

**WindowCovering**
TargetPositionLiftPercent100ths (added as a diagnostic entity, converted to the % that the shade is open to match the way in which the open/close slider presents the info.)

**OccupancySensing**
PIROccupiedToUnoccupiedDelay (added as a Number entity)


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

jvmahon committed:
Add Matter attributes

jvmahon committed:
Add Matter attributes

jvmahon committed:
Add Matter attributes

jvmahon committed:
Add Matter attributes

NoRi2909 commented on homeassistant/components/matter/strings.json:
```diff
@@ -220,6 +226,15 @@
         "name": "Spin speed"
+      },
+      "local_config_disabled": {
+        "name": "Local Configuration"
```
All those new strings need to be sentence-cased -> "Local configuration", "One-touch locking", "Sound volume", " Min PIN code length", "Max PIN code length", " Target open-to position"

jvmahon replied:
Fixed. Thanks.

jvmahon committed:
Update strings.json

marcelveldt commented on homeassistant/components/matter/select.py:
```diff
@@ -435,4 +435,65 @@ def _update_from_device(self) -> None:
+            key="LocalConfigDisabled",
+            entity_category=EntityCategory.CONFIG,
+            translation_key="local_config_disabled",
+            options=["Enable Local Configuration", "Disable Local Configuration"],
```
```suggestion
            options=["enable", "disable"],
```
Use strings.json to map this to human readable name.
That said: why is this a select entity and not a switch if its just a toggle ?

marcelveldt commented on homeassistant/components/matter/select.py:
```diff
@@ -435,4 +435,65 @@ def _update_from_device(self) -> None:
+            translation_key="local_config_disabled",
+            options=["Enable Local Configuration", "Disable Local Configuration"],
+            measurement_to_ha={
+                False: "Enable Local Configuration",
+                True: "Disable Local Configuration",
```
same comment as above

jvmahon replied:
Removed these lines. Will add back in a separate PR converting these to switch entities.

marcelveldt commented on homeassistant/components/matter/select.py:
```diff
@@ -435,4 +435,65 @@ def _update_from_device(self) -> None:
+                True: "Disable Local Configuration",
+            }.get,
+            ha_to_native_value={
+                "Enable Local Configuration": False,
+                "Disable Local Configuration": True,
```
same comment as above

jvmahon replied:
Removed these lines. Will add back in a separate PR converting these to switch entities.

marcelveldt commented on homeassistant/components/matter/select.py:
```diff
@@ -435,4 +435,65 @@ def _update_from_device(self) -> None:
+            key="EnableOneTouchLocking",
+            entity_category=EntityCategory.CONFIG,
+            translation_key="enable_one_touch_locking",
+            options=["Enabled", "Disabled"],
```
same comment as above

jvmahon replied:
Removed these lines. Will add back in a separate PR converting these to switch entities.

jvmahon committed:
Update homeassistant/components/matter/select.py

Co-authored-by: Marcel van der Veldt <m.vanderveldt@outlook.com>

jvmahon commented:
What's your preferred approach here - in the past, you've sometimes found it easier to make the suggested changes, but I can go back and make the suggested changes (chaging the select entities to switch entities, and providing the "strings" info). Whatever works best for you.

frenck commented:
Marking this PR as a draft, as there is a merge conflict.

Could you take a look @jvmahon?

../Frenck

jvmahon committed:
Update select.py

Deleted items to be added as switch entities instead.

jvmahon committed:
Update strings.json

jvmahon committed:
Update select.py

jvmahon committed:
Update strings.json

jvmahon commented:
@marcelveldt I think all the issues you flagged in this are now fixed.
I removed a number of attributes where you suggested they would be better represented as "switch" entities. I'll add those back in a separate PR.

joostlek committed:
Merge branch 'dev' into Add-Matter-Attributes

# Conflicts:
#	homeassistant/components/matter/sensor.py
#	homeassistant/components/matter/strings.json

joostlek committed:
Fix

NoRi2909 commented on homeassistant/components/matter/strings.json:
```diff
@@ -341,6 +356,15 @@
+        "name": "Max PIN code length"
+      },
+      "window_covering_target_position": {
+        "name": "Target open-to position"
```
I have a problem with "open-to", that word combination is nowhere to be found and confuses the machine translators, too.

Can we reword this in some way so it gets a definite meaning and translation?

jvmahon replied:
I changed it to "Target open position" which should resolve this concern.

NoRi2909 replied:
Hmm, isn't this the "Target _opening_ position" in percent?

The definition in the Matter specs reads:
"Indicates the position where the window covering lift will go or is moving to as a percentage"

This can be any position, not just one that we would call "open" (as the opposite of "closed").

jvmahon replied:
OK, changed to "opening"

jvmahon committed:
Update strings.json

jvmahon committed:
Update strings.json

joostlek committed:
Fix

MartinHjelmare commented on homeassistant/components/matter/select.py:
```diff
@@ -436,4 +436,27 @@ def _update_from_device(self) -> None:
+                "silent": 0,
+                "low": 1,
+                "medium": 3,
+                "high": 2,
```
This should probably correspond to line 450, right? Either change here or there.
```suggestion
                "high": 4,
```

jvmahon replied:
Line 450 was the mistake - Matter App. Cluster section 5.2.6.23 says the "high" value is 2. Thanks for catching this.

jvmahon committed:
Update select.py

jvmahon commented on homeassistant/components/matter/select.py:
```diff
@@ -436,4 +436,27 @@ def _update_from_device(self) -> None:
+                "high": 2,
+            }.get,
+        ),
+        entity_class=MatterAttributeSelectEntity,
```
Are you asking me to add this test, or was this for the "larger" HA team to handle at some future time? I'm assuming this not is for the other HA team members (if it was for me, I don't know how to set up test).
Otherwise, is this now ready to be integrated into the "main" code, or is there anything else you are asking me to do. 
thanks.

MartinHjelmare replied:
I think we should add a test, in a separate PR, before merging this PR. You don't have to do that.

joostlek replied:
#145440

MartinHjelmare reviewed (approved):
Thanks!

marcelveldt commented:
@jvmahon sorry for the late feedback. I just noticed that this PR brings in some additional entities which do not belong in the HA state entity:

MinPINCodeLength, MaxPINCodeLength and probably also TargetPositionLiftPercent100ths (I need to double check that).
That is just static information - there is no use for it as entity at all and it will only cause "noise" towards users in the UI.

We will remove those again in a follow-up PR.
In the future there might be a usecase to knowing this info while setting up a lock code in the HA interface but that is a different discussion and needs a proper architectural discussion first. Adding static info to the state machien is simply not allowed, hence we will reverse it.

Thanks for understanding.

MartinHjelmare commented on homeassistant/components/matter/sensor.py:
```diff
@@ -948,6 +948,42 @@ def _update_from_device(self) -> None:
+        entity_class=MatterSensor,
+        required_attributes=(clusters.DoorLock.Attributes.MaxPINCodeLength,),
+    ),
+    MatterDiscoverySchema(
+        platform=Platform.SENSOR,
+        entity_description=MatterSensorEntityDescription(
+            key="TargetPositionLiftPercent100ths",
+            entity_category=EntityCategory.DIAGNOSTIC,
+            translation_key="window_covering_target_position",
+            measurement_to_ha=lambda x: round((10000 - x) / 100),
+            native_unit_of_measurement=PERCENTAGE,
+        ),
+        entity_class=MatterSensor,
+        required_attributes=(
+            clusters.WindowCovering.Attributes.TargetPositionLiftPercent100ths,
+        ),
+    ),
```
https://github.com/home-assistant/core/pull/145713
